### PR TITLE
[IoT] Increase SDK and API version

### LINKED ISSUE #22257: `az iot dps linked-hub create` hub and DPS on different subscriptions broken!?
When running `az iot dps linked-hub create` from the az CLI it would appear that we are unable to specify linking to a hub in a different subscription from the DPS.

When running under the context being set to the subscription where the DPS exists, and despite passing a valid IoTHub connectionstring param to the command, it still responds with _"No IoT Hub found with name 'iot-hub-foo' in current subscription."_.

```
az iot dps linked-hub create --dps-name $IoTDPSName -g $ResourceGroupIoTDPS --connection-string $IoTHubCN
```

It's worth noting that the command obviously works when both the DPS and the IotHub exist in the same Azure Subscription. Additionally, I've also ensured that I'm owner of the resource in case it was possibly related to [this issue](https://github.com/Azure/azure-cli/issues/22038) for a related command.

Can anyone from the Azure team please confirm that my observations above are correct and we're currently unable to link an IoTHub to a DPS in a different Subscription?

### PR DESCRIPTION
**Pull request by vilit1**
**Related command**

`az iot hub`
`az iot hub create`
`az iot hub route test`
`az iot dps linked-hub create`

**Description**
Increase the API version to 2022-04-30-preview and pin the min SDK to azure-mgmt-iothub==2.3.0. This will enable customers to modify the control plane of hubs that have features from the 2022-04-30-preview API, specifically the Cosmos DB endpoints.

The Cosmos DB Endpoint feature will be released in the azure iot extension.

There is also some small fixes:
1. the fileupload_storage_container_uri parameter is no longer used and is removed in this pr. This will remove the warning message below that occurs during hub creation.
image
2. the app_properties and system_properties for hub routes will now be converted to a dictionary (and thus make the routes test work)
3. the dps linked hub will now have an error message about location - so that if a hub is provided in a different subscription, then the user will be prompted to give the location. This will fix the customer confusion as seen in this issue: https://github.com/Azure/azure-cli/issues/22257
4. Add preview to device streams as part of https://github.com/Azure/azure-cli/issues/24554
5. Fix a networking test in which `az iot device-update` was renamed to `az iot du`
6. Add no wait to iot hub create and delete

**Testing Guide**

Regenerated all affected iot commands. Also added a test to route test (that would have failed in the older version)

**History Notes**

[IoT] Fix #22257: `az iot dps linked-hub create`: Improve error handling for linked hubs
[IoT] `az iot hub create/delete`: Add `--no-wait` parameter to support no wait operation

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
IoT Hub

File: src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml
Comment (evelyn-ys) [inline src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml:71]: Please fix the test failure with new API version instead of manually replacing API verison in yaml files

Comment (vilit1) [reply src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml:71]: tracking issue here https://github.com/Azure/azure-cli/issues/24382

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml:71]: @vilit1 I believe this issue will also affect users. Before you solve this issue, we can't merge this PR, right?

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml:71]: @vilit1 Since we will launch the release of this sprint tomorrow, please resolve all the comments before tomorrow, otherwise this PR will not be able to catch up with the release of this sprint, and it can only be postponed to the next sprint **(12-06)**

Comment (vilit1) [reply src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml:71]: This will fix will be postponed until the fix is avaliable in the second week of november.

Comment (vilit1) [reply src/azure-cli/azure/cli/command_modules/security/tests/latest/recordings/test_security_iot.yaml:71]: Fixed!

File: src/azure-cli/azure/cli/command_modules/iot/tests/latest/test_iot_dps_commands.py
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/iot/tests/latest/test_iot_dps_commands.py:265]: Why delete these commands in test?

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py
Comment (zhoxing-ms) [inline src/azure-cli-core/azure/cli/core/profiles/_shared.py:246]: Since there is no functional upgrade in this PR, may I ask why we need to upgrade the api-version to the new preview version?

File: src/azure-cli/azure/cli/command_modules/iot/tests/latest/test_iot_dps_commands.py
Comment (vilit1) [reply src/azure-cli/azure/cli/command_modules/iot/tests/latest/test_iot_dps_commands.py:265]: These commands do not work - as the key in the connection string is case sensitive. I have checked with the service and this behavior is expected.

File: src/azure-cli-core/azure/cli/core/profiles/_shared.py
Comment (vilit1) [reply src/azure-cli-core/azure/cli/core/profiles/_shared.py:246]: This is the newer api and the service (and portal) will be using this as the main api. The newer api (and sdk) provide cosmos db endpoints and hub certificate migration functionality.
The changes will be in the extension that will rely on this newer api. See https://github.com/Azure/azure-iot-cli-extension/pull/556 and https://github.com/Azure/azure-iot-cli-extension/pull/547

Having one api will increase consistency and reduce confusion among customers since they will not see hub certificate properties or cosmos db endpoints with the older api.

Comment (digimaun) [reply src/azure-cli-core/azure/cli/core/profiles/_shared.py:246]: On top of what @vilit1 said, some background...

Historically for IoT _most_ control plane functionality resides in Azure CLI and all data plane resides in the IoT extension. We are aiming to pull in more control plane to the extension, starting with upcoming root cert migration & cosmos db endpoint capability. Currently the extension relies on the IoT Hub control plane/mgmt SDK provided by Az CLI, so we need to increment it to move forward.

Comment (zhoxing-ms) [reply src/azure-cli-core/azure/cli/core/profiles/_shared.py:246]: Got it, thanks~

File: src/azure-cli/azure/cli/command_modules/iot/commands.py
Comment (zhoxing-ms) [inline src/azure-cli/azure/cli/command_modules/iot/commands.py:187]: It is not recommended to downgrade GA commands to preview state. I'd like to confirm with you that the reason for changing it to preview is because the REST service contains new instability factors after using the new preview api-version, right?

**Comment by zhoxing-ms** [comment]
> [IoT Hub] Increase min SDK and API version
[IoT Hub] Remove warning during Hub creation and fix routes test
[IoT Dps] Improve error handling for linked hubs

We usually only write the changes that can be perceived by users in the history, so I personally think the first and second points can be deleted.
Could you describe the third point in more detail?

File: src/azure-cli/azure/cli/command_modules/iot/commands.py
Comment (vilit1) [reply src/azure-cli/azure/cli/command_modules/iot/commands.py:187]: The api has never been been in GA state as far as I recall. This is also an ask https://github.com/Azure/azure-cli/issues/24554

**Comment by vilit1** [comment]
> > [IoT Hub] Increase min SDK and API version
> > [IoT Hub] Remove warning during Hub creation and fix routes test
> > [IoT Dps] Improve error handling for linked hubs
> 
> We usually only write the changes that can be perceived by users and affect users' business in the history notes, so I personally think the first and second points can be deleted. Could you describe the third point in more detail?

the dps linked hub will now have an error message about location - so that if a hub is provided in a different subscription, then the user will be prompted to give the location. This will fix the customer confusion as seen in this issue: https://github.com/Azure/azure-cli/issues/22257

File: src/azure-cli/azure/cli/command_modules/iot/commands.py
Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/iot/commands.py:188]: OK, so this command was incorrectly declared GA states before, which was not in line with the expectation, right?

**Comment by zhoxing-ms** [comment]
> The second point will affect users too

In fact, the warning log has no real impact on the user's business, so we usually do not write the changes of warning log into the history notes.

> [IoT] Fix https://github.com/Azure/azure-cli/issues/22257: az iot dps linked-hub create: Improve error handling for linked hubs

This is my updated history note, please see if it meets your expectations

File: src/azure-cli/azure/cli/command_modules/iot/commands.py
Comment (digimaun) [reply src/azure-cli/azure/cli/command_modules/iot/commands.py:188]: Yea, the mistake was not classifying it as preview when it was released originally.

Comment (zhoxing-ms) [reply src/azure-cli/azure/cli/command_modules/iot/commands.py:188]: OK, got it.

**Review by zhoxing-ms — APPROVED** [review]

**Review by evelyn-ys — APPROVED** [review]

**Review by necusjz — APPROVED** [review]

**Pull request merged**
